compiler/src/dotty/tools/dotc/core/Scopes.scala

```diff
@@ -163,13 +165,13 @@ object Scopes {
      *  a copy with the matching symbols.
      */
     final def filteredScope(p: Symbol => Boolean)(using Context): Scope = {
-      var result: MutableScope = null
+      var result: MutableScope | Null = null
       for (sym <- iterator)
         if (!p(sym)) {
           if (result == null) result = cloneScope
-          result.unlink(sym)
+          result.nn.unlink(sym)
         }
-      if (result == null) this else result
+      if (result == null) this else result.nn
```

Comment:
Is it because of the for loop that we can't? Does the for loop body turn into a closure, and then inference fails because the closure captures `result`?